File: with_laurent/tail-recursion-optimization.js
// def plus(a, b):
//     if a == 0:
//         return b
//     return plus(a - 1, b + 1)

function plus(a, b) {
    if (a === 0) {
        return b
    }
    return plus(a - 1, b + 1)
}

// console.log(plus(4, 5))
plus(1000000, 0) // RangeError: Maximum call stack size exceeded
// => No tail recursion elimination in JS (with node interpreter)
// NB: apparently, some interpreters do perform tail-recursion elimination

File: with_laurent/tail-recursion-optimization.py
def plus(a, b):
    if a == 0:
        return b
    return plus(a - 1, b + 1)


# print(plus(4, 5))
plus(1000000, 0)  # RecursionError: maximum recursion depth exceeded in comparison

# => No tail recursion elimination in Python.
# Also, explanation by Guido Van Rossum about this is here:
# https://neopythonic.blogspot.com/2009/04/tail-recursion-elimination.html
